google-cloud-spanner/src/main/java/com/google/cloud/spanner/Mutation.java

```diff
@@ -464,7 +494,44 @@ static void toProto(Iterable<Mutation> mutations, List<com.google.spanner.v1.Mut
     }
     // Flush last item.
     if (proto != null) {
+      com.google.spanner.v1.Mutation builtMutation = proto.build();
       out.add(proto.build());
+      if (checkIfInsertMutationWithLargeValue(builtMutation, largestInsertMutation)) {
+        largestInsertMutation = builtMutation;
+      }
+      addMutationsExcludingInsert(builtMutation, allMutationsExcludingInsert);
+    }
+
+    // Select a random mutation based on the heuristic.
+    if (!allMutationsExcludingInsert.isEmpty()) {
+      return allMutationsExcludingInsert.get(
+          ThreadLocalRandom.current().nextInt(allMutationsExcludingInsert.size()));
+    } else {
+      return largestInsertMutation;
+    }
+  }
+
+  // Returns true if the input mutation is of type INSERT and has more values than the current
+  // largest insert mutation.
+  private static boolean checkIfInsertMutationWithLargeValue(
+      com.google.spanner.v1.Mutation mutation,
+      com.google.spanner.v1.Mutation largestInsertMutation) {
+    // If largestInsertMutation is a default instance of Mutation, replace it with the current
+    // INSERT mutation, even if it contains zero values.
+    if (!largestInsertMutation.hasInsert()) {
```

Comment:
I think that this should be:

```suggestion
    if (mutation.hasInsert() && !largestInsertMutation.hasInsert()) {
```

Please add a test that fails with the current implementation and succeeds with the above change (unless I missed something here, and the current implementation is correct).

Reply:
I just realized that this won't fail, as there are only two possibilities:
1. There are only insert mutations. Then the largest insert mutation will be used.
2. There is at least one non-insert mutation. Although it is possible that that mutation is set as 'the largest insert mutation' with the current if statement, it would not really make any difference, as the largest insert mutation won't be used anyways.

We should still fix the above if condition, though, as the current implementation is a bit confusing.

There is however another (small) optimization possible here; Once `allMutationsExcludingInserts.isEmpty()` returns false, you can skip the tracking of the largest insert mutation all-together, as it won't be used anyways.

Reply:
1. Thanks for pointing this out. It does not affect the result, but it is indeed a flaw. Updated the code.
2. Added the optimization.